pg_analytics/src/datafusion/datatype.rs

```diff
@@ -234,136 +236,146 @@ impl DatafusionMapProducer {
         (datum, is_null)
     }
 
-    pub fn array(
+    pub fn array_from_slots(
         sql_data_type: SQLDataType,
         slots: *mut *mut pg_sys::TupleTableSlot,
         nslots: usize,
         col_idx: usize,
+    ) -> Result<ArrayRef, ParadeError> {
+        let mut vec = Vec::with_capacity(nslots);
+        for row_idx in 0..nslots {
+            let (datum, is_null) = unsafe { Self::tuple_info(slots, row_idx, col_idx) };
+            if is_null {
+                vec.push(None);
+            } else {
+                unsafe { vec.push(Some(*datum)) };
+            }
+        }
+
+        Self::array(sql_data_type, vec)
+    }
+
+    pub fn array(
+        sql_data_type: SQLDataType,
+        datums: Vec<Option<Datum>>,
     ) -> Result<ArrayRef, ParadeError> {
         let datafusion_type = DatafusionTypeTranslator::from_sql_data_type(sql_data_type)?;
 
         match datafusion_type {
             DataType::Boolean => {
-                let mut vec = Vec::with_capacity(nslots);
-
-                for row_idx in 0..nslots {
-                    let (datum, is_null) = unsafe { Self::tuple_info(slots, row_idx, col_idx) };
-                    if is_null {
-                        vec.push(None);
-                    } else {
-                        vec.push(unsafe { bool::from_datum(*datum, false) })
+                let mut vec = Vec::with_capacity(datums.len());
+
+                for datum in datums.iter() {
+                    match datum {
+                        Some(datum) => vec.push(unsafe { bool::from_datum(*datum, false) }),
+                        None => vec.push(None)
                     }
                 }
+
                 Ok(Arc::new(BooleanArray::from(vec)))
             }
             DataType::Utf8 => {
-                let mut vec = Vec::with_capacity(nslots);
-
-                for row_idx in 0..nslots {
-                    let (datum, is_null) = unsafe { Self::tuple_info(slots, row_idx, col_idx) };
-                    if is_null {
-                        vec.push(None);
-                    } else {
-                        vec.push(unsafe { String::from_datum(*datum, false) })
+                let mut vec = Vec::with_capacity(datums.len());
+
+                for datum in datums.iter() {
+                    match datum {
```

Comment:
Have you tested if this works with `null` values? I'm not 100% sure if datum is `None` if it's null, I've usually seen that info coming from the `is_null` LIst.